src-tauri/src/terminal.rs

```diff
@@ -0,0 +1,414 @@
+use crate::steam::SteamDownload;
+use async_process::Command;
+use serde::Serialize;
+use std::{env, fs};
+use std::fs::File;
+use std::os::unix::fs::PermissionsExt;
+use crate::get_os;
+
+/// Represents a terminal that can be used to run commands.
+/// **Should be in sync with the terminal dropdown in the frontend.**
+#[derive(Debug, Serialize, PartialEq)]
+pub enum Terminal {
+    GNOMETerminal,
+    Alacritty,
+    Konsole,
+    GNOMEConsole,
+    Xfce4Terminal,
+    DeepinTerminal,
+    Terminator,
+    Terminology,
+    Kitty,
+    LXTerminal,
+    Tilix,
+    CoolRetroTerm,
+    XTerm,
+    CMD,
+    Terminal
+}
+
+
+impl Terminal {
+    /// Iterates through each terminal
+    pub fn iter() -> impl Iterator<Item=Terminal> {
+        use self::Terminal::*;
+
+        vec![
+            GNOMETerminal, Alacritty, Konsole, GNOMEConsole, Xfce4Terminal, DeepinTerminal, Terminator, Terminology, Kitty, LXTerminal, Tilix, CoolRetroTerm, XTerm, CMD, Terminal
+        ].into_iter()
+    }
+
+    /// Get terminal from index in order of the [`Terminal`] enum
+    pub fn from_index(index: &u8) -> Option<Terminal> {
+        Terminal::iter().nth(*index as usize)
+    }
+
+    /// Get the index of a terminal in the order of the [`Terminal`] enum
+    /// Returns `None` if the terminal is not found.
+    pub fn index(&self) -> Option<u8> {
+        Terminal::iter().position(|x| x == *self).map(|x| x as u8)
+    }
+
+
+    /// Get total number of terminals **possible** depending on the OS
+    pub fn total() -> u8 {
+        if get_os() != "windows" || get_os() == "macos" {
+            return 1;
+        }
+
+        Terminal::iter().count() as u8 - 1 // -1 because cmd is not available on linux
+    }
+
+    /// Get the pretty name of a terminal
+    pub fn pretty_name(&self) -> &str {
+        match self {
+            Terminal::GNOMETerminal => "GNOME Terminal",
+            Terminal::GNOMEConsole => "GNOME Console",
+            Terminal::Konsole => "Konsole",
+            Terminal::Xfce4Terminal => "Xfce Terminal",
+            Terminal::Terminator => "Terminator",
+            Terminal::Terminology => "Terminology",
+            Terminal::XTerm => "XTerm",
+            Terminal::Kitty => "Kitty",
+            Terminal::LXTerminal => "LXTerminal",
+            Terminal::Tilix => "Tilix",
+            Terminal::DeepinTerminal => "Deepin Terminal",
+            Terminal::CoolRetroTerm => "cool-retro-term",
+            Terminal::Alacritty => "Alacritty",
+            Terminal::CMD => "cmd",
+            Terminal::Terminal => "Terminal"
+        }
+    }
+
+
+    //region Probing a terminal
+    /// Checks if a [`Terminal`] is installed.
+    /// **See:** [`get_installed_terminals`]
+    pub async fn installed(&self) -> bool {
```

Comment:
this looks like it could be a lot more compact and less repeating. 
but making it one would probably be more complex cause of all the exceptions and different naming/run cmd

Reply:
Problem is it varies between terminals. But like you mentioned it could be done with a JSON file, and then serializing it to a field in the `Terminal` struct but that would probably make the problem, worse because you still need to add terminals to multiple places in the codebase.